### PR TITLE
cop.h - add support for refcounted cop_file under threads and cop_warnings always.

### PR DESCRIPTION
**Pull request by demerphq**
We have a weird bifurcation of the cop logic around threads. With threads we use a char * cop_file member, without it we use a GV * and replace cop_file with cop_filegv.

The GV * code refcounts filenames and more or less efficiently shares the filename amongst many opcodes. However under threads we were simplify copying the filenames into each opcode. This is because in theory opcodes created in one thread can be destroyed in another. I say in theory because as far as I know the core code does not actually do this. But we have tests that you can construct a perl, clone it, and then destroy the original, and have the copy work just fine, this means that opcodes constructed in the main thread will be destroyed in the cloned thread. This in turn means that you can't put SV derived structures into the op-tree under threads. Which is why we can not use the GV * stategy under threads.

As such this code adds a new struct/type RCPV, which is a refcounted string using shared memory. This is implemented in such a way that code that previously used a char * can continue to do so, as the refcounting data is located a specific offset before the char * pointer itself. This also allows the len data to embedded "into" the PV, which allows us to expose macros to acces the length of what is in theory a null terminated string.

    struct rcpv {
        UV      refcount;
        STRLEN  len;
        char    pv[1];
    };
    typedef struct rcpv RCPV;

The struct is sized appropriately on creation in rcpv_new() so that the pv member contains the full string plus a null byte. It then returns a pointer to the pv member of the struct. Thus the refcount and length and embedded at a predictable offset in front of the char *, which means we do not have to change any types for members using this.

We provide three operations: rcpv_new(), rcpv_copy() and rcpv_free(), which roughly correspond with newSVpv(), SvREFCNT_inc(), SvREFCNT_dec(), and a handful of macros as well. We also expose SAVERCPVFREE which is similar to SAVEGENERICSV but operates on pv's constructed with rcpv_new().

Currently I have not restricted use of this logic to threaded perls. We simply do not use it in unthreaded perls, but I see no reason we couldn't normalize the code to use this in both cases, except possibly that actually the GV case is more efficient.

Note that rcpv_new() does NOT use a hash table to dedup strings. Two calls to rcpv_new() with the same arguments will produce two distinct pointers with their own refcount data.

Refcounting the cop_file data was Tony Cook's idea.

**ADDENDUM:** It turned out to be very simple to convert cop_warnings to use the same rcpv infrastructure under the hood, so this PR also includes cop_warnings as well.

### CODE REVIEW

**Comment by demerphq** [comment]
@tonycoz this is for you.

File: op.c
Comment (tonycoz) [inline op.c]: The NULL check belongs before the accesses.

File: cop.h
Comment (tonycoz) [inline cop.h]: `UV refcount` makes the reference count 64-bit on a 32-bit platform with `-Duse64bitint`, it could be STRLEN instead.

**Comment by tonycoz** [comment]
Why not just put the RCPV pointer into cop_file? (probably renamed)

File: op.c
Comment (demerphq) [reply op.c]: Willfix.

File: cop.h
Comment (demerphq) [reply cop.h]: willfix

**Comment by demerphq** [comment]
> Why not just put the RCPV pointer into cop_file? (probably renamed)

The main reason is that it would require more code to change than I think is necessary or reasonable. For instance in scope.c we can use SAVEPPTR with a RCPV pv without needing to create a bespoke SAVEt_ for RCPV *. The scope.c logic and headers are all manually maintained, so adding new stuff is highly error prone and requires modifying a lot of comments and code. Ditto for the B code, doing the change like this meant I didn't need to change B at all. That seems like a good thing.

The other point is more general, I have a feeling this logic could long term be used in places where we store char * pointers and where it would be awkward to change their type. 

Also I want to look into using them for warning strings too, and i think there it will be awkward to change type. 

I am not /against/ using the type directly, but I feel this is a time where being parsimonious with the amount of change is a good thing.

**Comment by demerphq** [comment]
BTW, i will rebase squash a bunch of these patches once review i complete.

**Comment by demerphq** [comment]
@tonycoz it now includes the cop_warnings stuff. I might do more work on that soon, but it demonstrates how the pv API is helpful. It is not perfect, but it works.

File: cop.h
Comment (tonycoz) [inline cop.h]: The %6d could be %6zu and avoid the cast to int for RCPV_REFCOUNT().

File: op.h
Comment (tonycoz) [inline op.h]: mutext -> mutex

File: sv.c
Comment (tonycoz) [inline sv.c:15143]: Should this OP_REFCNT_LOCK?

I see what you mean about the (probably wild) pointer issue, perhaps this should croak instead.

A real fix would require tracking every object across the clone, and keeping the object base pointer and the offset to the pointer we're working on in the save record.  That probably isn't practical.

**Comment by tonycoz** [comment]
> I am not /against/ using the type directly, but I feel this is a time where being parsimonious with the amount of change is a good thing.

I'm mostly concerned with downstream code that tries to access cop_file/cop_warnings directly, though from looking today, there isn't much doing that.

Devel::Size will silently but mostly harmlessly not count the space needed for the size/refcount fields for cop_file, though it does appear to handle the value being shared across COPs correctly.

Syntax::Keyword::Try appears to access cop_warnings safely.

Please consider this concern resolved.

File: cop.h
Comment (demerphq) [reply cop.h]: Ah, thanks, will fix.

File: sv.c
Comment (demerphq) [reply sv.c:15143]: > Should this OP_REFCNT_LOCK?

It locks in rcpv_copy when it does the refcount increment. Is that sufficient you think?

> I see what you mean about the (probably wild) pointer issue, perhaps this should croak instead.

I might try it just to see if anything blows up. I kinda wonder about a number of the cases in that code, it seems like several of them would have this problem.

Comment (tonycoz) [reply sv.c:15143]: > > Should this OP_REFCNT_LOCK?
> 
> It locks in rcpv_copy when it does the refcount increment. Is that sufficient you think?

Doh, of course.  Please ignore me on that then.

**Comment by demerphq** [comment]
@nwc10 - regarding Devel::Size, is the repo handy anywhere so i can patch it to handle the new RCPVs?

**Review by tonycoz — APPROVED** [review]

**Comment by demerphq** [comment]
@tonycoz thanks - i have squashed the fixups, and ill let CI run for a while before i merge.

**Pull request merged**
